Add aria-invalid attribute to all required fields (LG-3531)

zachmargolis opened the pull request:
I opted to enforce this in our accessibility specs by adding my own custom RSpec matcher.

Another option would be to upgrade axe-matchers (which has been replaced)

zachmargolis commented on app/views/devise/passwords/new.html.erb:
```diff
@@ -14,7 +14,7 @@
                        url: user_password_path,
                        html: { autocomplete: 'off', method: :post, role: 'form' }) do |f| %>
-  <%= f.input :email, required: true, input_html: { 'aria-describedby': 'email-description' } %>
+  <%= f.input :email, required: true, input_html: { aria: { invalid: true, describedby: 'email-description' } } %>
```
For some reason, doing `'aria-invalid': true` just never rendered to the DOM, so I had to do this syntax, similar to how `data-` prefixed keys work

aduth replied:
Curious if `'aria-invalid': 'true'` would have worked (i.e. maybe boolean values only treated specially for `aria` and `data` groupings?).

zachmargolis replied:
No dice! That did not seem to work. I have a *feeling* that this is due to how the `input_html` is passed through various translation layers to ActionView, but haven't dug deep enough to confirm

aduth replied:
🤷  IMO, the object form seems quite nice as it is anyhow.

aduth commented:
From related resources, I think we want the default value to be `aria-invalid="false"`, and then switch it to `true` if and when we want to signal the field is invalid.

>To stop form controls from announcing as invalid by default, one can add `aria-invalid="false"` to any necessary element. [...] When its time to mark a form control as invalid, the attribute’s value should be set to “true”. Ideally, a control should be set to invalid only after a required control has been focused, and then blurred, without a proper string being entered.

https://developer.paciellogroup.com/blog/2019/02/required-attribute-requirements/

Guidance around when to trigger the switch is mixed, but blur seems to be a reasonable default:

>Note that it is not necessary to validate the fields immediately on blur; the application could wait until the form is submitted (though this is not necessarily recommended).

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-invalid_attribute#Example_1_Simple_form_validation

In our case, we might be able to handle this as part of the existing [`form-validation.js`](https://github.com/18F/identity-idp/blob/master/app/javascript/packs/form-validation.js) logic. This only applies to fields with the `field` class, but that should cover most of the `validated_form_for` usage.

zachmargolis commented:
> From related resources, I think we want the default value to be `aria-invalid="false"`, and then switch it to `true` if and when we want to signal the field is invalid.

Ah, whoops, thanks!

> In our case, we might be able to handle this as part of the existing [`form-validation.js`](https://github.com/18F/identity-idp/blob/master/app/javascript/packs/form-validation.js) logic. This only applies to fields with the `field` class, but that should cover most of the `validated_form_for` usage.

Good point -- I'll take a shot at adding the logic there

zachmargolis commented:
> > From related resources, I think we want the default value to be `aria-invalid="false"`, and then switch it to `true` if and when we want to signal the field is invalid.
> 
> Ah, whoops, thanks!

Updated in 069dc7f0ec94d885b45899f3d9b6cf840c387a80

> > In our case, we might be able to handle this as part of the existing [`form-validation.js`](https://github.com/18F/identity-idp/blob/master/app/javascript/packs/form-validation.js) logic. This only applies to fields with the `field` class, but that should cover most of the `validated_form_for` usage.
> 
> Good point -- I'll take a shot at adding the logic there

Added in 4d75bc42f467acde15900d26e4a0973725a552c7

| before | after |
| --- | ---- |
| Blank form is invalid before it is interacted with <img width="667" alt="Screen Shot 2020-10-16 at 9 58 56 AM" src="https://user-images.githubusercontent.com/458784/96287348-a56fa300-0f96-11eb-8077-5a28328f2281.png"> | Blank form is just "required" not "invalid"  <img width="712" alt="Screen Shot 2020-10-16 at 9 59 49 AM" src="https://user-images.githubusercontent.com/458784/96287378-b15b6500-0f96-11eb-8ee3-d0bbedd69abe.png"> |
| After form labeled as invalid after interacted with (**expected**) <img width="861" alt="Screen Shot 2020-10-16 at 10 03 44 AM" src="https://user-images.githubusercontent.com/458784/96287517-eff11f80-0f96-11eb-8400-dec5fe6cf372.png"> | After interacting, invalid form still labeled as invalid  <img width="651" alt="Screen Shot 2020-10-16 at 9 58 41 AM" src="https://user-images.githubusercontent.com/458784/96287440-c89a5280-0f96-11eb-9d26-c3b7c0c8ef92.png"> |

aduth reviewed (approved):
Looks good 👍
